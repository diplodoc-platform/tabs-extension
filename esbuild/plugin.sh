esbuild src/plugin/index.ts --outfile=plugin/index.js \
  --bundle --platform=node \
  --sourcemap \
  --define:PACKAGE="\"$(node -pe "require('./package.json').name")\""
