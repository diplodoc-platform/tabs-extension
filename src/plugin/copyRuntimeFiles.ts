import {dirname, join, resolve} from 'node:path';
import {mkdirSync, copyFileSync} from 'node:fs';

const PATH_TO_RUNTIME = '../runtime';

export function copyRuntimeFiles(
    {
        runtimeJsPath,
        runtimeCssPath,
        output,
    }: {runtimeJsPath: string; runtimeCssPath: string; output: string},
    cache: Set<string>,
) {
    const runtimeFiles = {
        'index.js': runtimeJsPath,
        'yfm.css': runtimeCssPath,
    };
    for (const [originFile, outputFile] of Object.entries(runtimeFiles)) {
        const file = join(PATH_TO_RUNTIME, originFile);
        if (!cache.has(file)) {
            cache.add(file);
            copyFile(resolve(__dirname, file), join(output, outputFile));
        }
    }
}

function copyFile(from: string, to: string) {
    mkdirSync(dirname(to), {recursive: true});
    copyFileSync(from, to);
}
