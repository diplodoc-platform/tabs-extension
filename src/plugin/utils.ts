export function generateID() {
    const id = Math.random().toString(36).substr(2, 8);
    return id.substring(id.length - 8);
}

export function addHiddenProperty<
    B extends Record<string | symbol, unknown>,
    F extends string | symbol,
    V,
>(box: B, field: F, value: V) {
    if (!(field in box)) {
        Object.defineProperty(box, field, {
            enumerable: false,
            value: value,
        });
    }

    return box as B & {[P in F]: V};
}
