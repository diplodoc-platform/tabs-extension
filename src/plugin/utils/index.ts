export function addHiddenProperty<
    B extends Record<string | symbol, unknown>,
    F extends string | symbol,
    V,
>(box: B, field: F, value: V) {
    if (!(field in box)) {
        Object.defineProperty(box, field, {
            enumerable: false,
            value: value,
        });
    }

    return box as B & {[P in F]: V};
}

export * from './strings';
export * from './tabs';
export * from './files';
