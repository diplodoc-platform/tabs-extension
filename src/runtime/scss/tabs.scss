.yfm-tabs {
    margin-bottom: 15px;

    .yfm:not(.yfm_no-list-reset) & ol {
        counter-reset: tabs-list;

        & > li {
            counter-increment: tabs-list;

            &::before {
                content: counters(tabs-list, '.') '. ';
            }
        }
    }
}

.yfm-tab-list {
    display: flex;
    flex-wrap: wrap;
    margin-bottom: 15px;

    border-bottom: 1px solid rgba(0, 0, 0, 0.07);
}

.yfm-tab {
    margin-bottom: -1px;
    margin-right: 20px;
    border-bottom: 2px solid transparent;

    padding: 6px 0 7px 0;
    font-weight: 700;
    outline: none;
    cursor: pointer;
    user-select: none;
    white-space: break-spaces;

    &:last-child {
        margin-right: 0;
    }

    &:hover, &:active {
        color: #004080;
    }

    &.active {
        border-bottom-color: #027bf3;
    }

    &:focus {
        outline: 2px solid #4d90fe;
        border-radius: 2px;
    }
    &:focus:not(:focus-visible) {
        outline: 0;
    }
}

.yfm-tab-panel {
    visibility: hidden;
    overflow: hidden;
    height: 0;

    &:first-child {
        margin-top: 0 !important;
    }

    &:last-child {
        margin-bottom: 0 !important;
    }

    &.active {
        visibility: visible;
        height: auto;
    }
}
