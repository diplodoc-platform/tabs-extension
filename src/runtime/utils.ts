export const getEventTarget = (event: Event) => {
    const path = event.composedPath();
    return Array.isArray(path) && path.length > 0 ? path[0] : event.target;
};

export const isCustom = (event: Event) => {
    const target = getEventTarget(event);
    return !target || !(target as HTMLElement).matches;
};
