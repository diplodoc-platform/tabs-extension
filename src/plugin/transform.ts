import MarkdownIt from 'markdown-it';
import StateCore from 'markdown-it/lib/rules_core/state_core';
import Token from 'markdown-it/lib/token';
import type {MarkdownItPluginCb} from '@doc-tools/transform/lib/plugins/typings';

import {addHiddenProperty, generateID} from './utils';
import {copyRuntimeFiles} from './copyRuntimeFiles';
import {getName, getTabId, getTabKey} from './getTabId';
import {
    ACTIVE_CLASSNAME,
    DEFAULT_TABS_GROUP_PREFIX,
    GROUP_DATA_KEY,
    TAB_ACTIVA_KEY,
    TABS_CLASSNAME,
    TABS_LIST_CLASSNAME,
    TAB_CLASSNAME,
    TAB_DATA_ID,
    TAB_DATA_KEY,
    TAB_PANEL_CLASSNAME,
} from '../common';

export type PluginOptions = {
    runtimeJsPath: string;
    runtimeCssPath: string;
    containerClasses: string;
    bundle: boolean;
};

const TAB_RE = /`?{% list tabs( group=([^ ]*))? %}`?/;

let runsCounter = 0;

export type Tab = {
    name: string;
    tokens: Token[];
};

function findTabs(tokens: Token[], idx: number) {
    const tabs = [];
    let i = idx,
        nestedLevel = -1,
        pending: Tab = {name: '', tokens: []};

    while (i < tokens.length) {
        const token = tokens[i];

        switch (token.type) {
            case 'ordered_list_open':
            case 'bullet_list_open':
                if (nestedLevel > -1) {
                    pending.tokens.push(token);
                }
                nestedLevel++;
                break;
            case 'list_item_open':
                if (nestedLevel) {
                    pending.tokens.push(token);
                } else {
                    pending = {name: '', tokens: []};
                }
                break;
            case 'list_item_close':
                if (nestedLevel) {
                    pending.tokens.push(token);
                } else {
                    tabs.push(pending);
                }
                break;
            case 'ordered_list_close':
            case 'bullet_list_close':
                if (!nestedLevel) {
                    return {
                        tabs,
                        index: i,
                    };
                }

                nestedLevel--;

                pending.tokens.push(token);

                break;
            case 'paragraph_open':
                if (
                    !pending &&
                    tokens[i + 1].content &&
                    tokens[i + 1].content.trim() === '{% endlist %}'
                ) {
                    return {
                        tabs,
                        index: i + 2,
                    };
                }

                if (!pending.name && tokens[i + 1].type === 'inline') {
                    pending.name = tokens[i + 1].content;
                    i += 2;
                } else {
                    pending.tokens.push(token);
                }
                break;
            default:
                pending.tokens.push(token);
        }

        i++;
    }

    return {
        tabs,
        index: i,
    };
}

function insertTabs(
    tabs: Tab[],
    state: StateCore,
    {start, end}: {start: number; end: number},
    {
        containerClasses,
        tabsGroup,
        runId,
    }: {containerClasses: string; tabsGroup: string; runId: string},
) {
    const tabsTokens = [];
    const tabListTokens = [];
    const tabPanelsTokens = [];
    const tabsOpen = new state.Token('tabs_open', 'div', 1);
    const tabsClose = new state.Token('tabs_close', 'div', -1);
    const tabListOpen = new state.Token('tab-list_open', 'div', 1);
    const tabListClose = new state.Token('tab-list_close', 'div', -1);

    tabsOpen.block = true;
    tabsClose.block = true;
    tabListOpen.block = true;
    tabListClose.block = true;

    tabsOpen.attrSet('class', [TABS_CLASSNAME, containerClasses].filter(Boolean).join(' '));
    tabsOpen.attrSet(GROUP_DATA_KEY, tabsGroup);
    tabListOpen.attrSet('class', TABS_LIST_CLASSNAME);
    tabListOpen.attrSet('role', 'tablist');

    for (let i = 0; i < tabs.length; i++) {
        const tabOpen = new state.Token('tab_open', 'div', 1);
        const tabInline = new state.Token('inline', '', 0);
        const tabText = new state.Token('text', '', 0);
        const tabClose = new state.Token('tab_close', 'div', -1);
        const tabPanelOpen = new state.Token('tab-panel_open', 'div', 1);
        const tabPanelClose = new state.Token('tab-panel_close', 'div', -1);

        const tab = tabs[i];
        const tabId = getTabId(tab, {runId});
        const tabKey = getTabKey(tab);
        tab.name = getName(tab);

        const tabPanelId = generateID();

        tabText.content = tabs[i].name;
        tabInline.children = [tabText];
        tabOpen.block = true;
        tabClose.block = true;
        tabPanelOpen.block = true;
        tabPanelClose.block = true;
        tabOpen.attrSet(TAB_DATA_ID, tabId);
        tabOpen.attrSet(TAB_DATA_KEY, tabKey);
        tabOpen.attrSet(TAB_ACTIVA_KEY, i === 0 ? 'true' : 'false');
        tabOpen.attrSet('class', TAB_CLASSNAME);
        tabOpen.attrSet('role', 'tab');
        tabOpen.attrSet('aria-controls', tabPanelId);
        tabOpen.attrSet('aria-selected', 'false');
        tabOpen.attrSet('tabindex', '-1');
        tabPanelOpen.attrSet('id', tabPanelId);
        tabPanelOpen.attrSet('class', TAB_PANEL_CLASSNAME);
        tabPanelOpen.attrSet('role', 'tabpanel');
        tabPanelOpen.attrSet('aria-labelledby', tabId);
        tabPanelOpen.attrSet('data-title', tab.name);

        if (i === 0) {
            tabOpen.attrJoin('class', ACTIVE_CLASSNAME);
            tabOpen.attrSet('aria-selected', 'true');
            tabOpen.attrSet('tabindex', '0');
            tabPanelOpen.attrJoin('class', ACTIVE_CLASSNAME);
        }

        tabListTokens.push(tabOpen, tabInline, tabClose);
        tabPanelsTokens.push(tabPanelOpen, ...tabs[i].tokens, tabPanelClose);
    }

    tabsTokens.push(tabsOpen);
    tabsTokens.push(tabListOpen);
    tabsTokens.push(...tabListTokens);
    tabsTokens.push(tabListClose);
    tabsTokens.push(...tabPanelsTokens);
    tabsTokens.push(tabsClose);

    state.tokens.splice(start, end - start + 1, ...tabsTokens);

    return tabsTokens.length;
}

function findCloseTokenIdx(tokens: Token[], idx: number) {
    let level = 0;
    let i = idx;
    while (i < tokens.length) {
        if (matchOpenToken(tokens, i)) {
            level++;
        } else if (matchCloseToken(tokens, i)) {
            if (level === 0) {
                return i;
            }
            level--;
        }

        i++;
    }

    return null;
}

function matchCloseToken(tokens: Token[], i: number) {
    return (
        tokens[i].type === 'paragraph_open' &&
        tokens[i + 1].type === 'inline' &&
        tokens[i + 1].content.trim() === '{% endlist %}'
    );
}

function matchOpenToken(tokens: Token[], i: number) {
    return (
        tokens[i].type === 'paragraph_open' &&
        tokens[i + 1].type === 'inline' &&
        tokens[i + 1].content.match(TAB_RE)
    );
}

export function transform({
    runtimeJsPath = '_assets/tabs-extension.js',
    runtimeCssPath = '_assets/tabs-extension.css',
    containerClasses = '',
    bundle = true,
}: Partial<PluginOptions> = {}) {
    const tabs: MarkdownItPluginCb<{output: string}> = function (md: MarkdownIt, {output = '.'}) {
        const plugin = (state: StateCore) => {
            const {env, tokens} = state;
            const runId = String(++runsCounter);

            addHiddenProperty(env, 'bundled', new Set<string>());

            let i = 0;
            let tabsAreInserted = false;

            while (i < tokens.length) {
                const match = matchOpenToken(tokens, i);
                const openTag = match && match[0];
                const isNotEscaped = openTag && !(openTag.startsWith('`') && openTag.endsWith('`'));

                if (!match || !isNotEscaped) {
                    i++;
                    continue;
                }

                const closeTokenIdx = findCloseTokenIdx(tokens, i + 3);

                if (!closeTokenIdx) {
                    tokens[i].attrSet('YFM005', 'true');
                    i += 3;
                    continue;
                }

                const tabsGroup = match[2] || `${DEFAULT_TABS_GROUP_PREFIX}${generateID()}`;

                const {tabs, index} = findTabs(state.tokens, i + 3);

                if (tabs.length > 0) {
                    i += insertTabs(
                        tabs,
                        state,
                        {start: i, end: index + 3},
                        {
                            containerClasses,
                            tabsGroup,
                            runId,
                        },
                    );
                    tabsAreInserted = true;
                } else {
                    state.tokens.splice(i, index - i);
                }
            }

            if (tabsAreInserted) {
                env.meta = env.meta || {};
                env.meta.script = env.meta.script || [];
                env.meta.style = env.meta.style || [];
                env.meta.script.push(runtimeJsPath);
                env.meta.style.push(runtimeCssPath);

                if (bundle) {
                    copyRuntimeFiles({runtimeJsPath, runtimeCssPath, output}, env.bundled);
                }
            }
        };

        try {
            md.core.ruler.before('curly_attributes', 'tabs', plugin);
        } catch (e) {
            md.core.ruler.push('tabs', plugin);
        }
    };

    return tabs;
}
