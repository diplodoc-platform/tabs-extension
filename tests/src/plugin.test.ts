import type MarkdownIt from 'markdown-it';

import dedent from 'ts-dedent';
import transform from '@diplodoc/transform';

import * as tabsExtension from '../../src/plugin/transform';

import {callPlugin, tokenize} from './utils';
import {base, escaped, nestedTokenTypes, simpleTab, vertical} from './data/tabs';

const defaultContent = [
    '# Create a folder',
    '',
    '{% list tabs %}',
    '',
    '- Python',
    '',
    '  About python',
    '',
    '- Tab with list',
    '  - One',
    '  - Two',
    '',
    '- Tab with list',
    '  1. One',
    '  2. Two',
    '',
    '{% endlist %}',
    '',
    'After tabs',
];

const defaultVerticalContent = [
    '# Create a folder',
    '',
    '{% list tabs radio %}',
    '',
    '- Python',
    '',
    '  About python',
    '',
    '- Tab with list',
    '  - One',
    '  - Two',
    '',
    '- Tab with list',
    '  1. One',
    '  2. Two',
    '',
    '{% endlist %}',
    '',
    'After tabs',
];

const convertAttrsToObject = ({attrs}: MarkdownIt.Token) =>
    attrs?.reduce((acc: Record<string, string>, [name, value]) => {
        acc[name] = value;

        return acc;
    }, {}) || {};

function makeTransform(params?: {
    transformOptions?: Partial<tabsExtension.PluginOptions>;
    content?: string[];
}) {
    return callPlugin(
        tabsExtension.transform({
            bundle: false,
            features: {enabledVariants: {radio: true, regular: true}},
            ...params?.transformOptions,
        }),
        tokenize(params?.content || defaultContent),
    );
}

function html(text: string, opts?: tabsExtension.PluginOptions) {
    const {result} = transform(text, {
        needToSanitizeHtml: false,
        plugins: [
            tabsExtension.transform({
                bundle: false,
                features: {enabledVariants: {radio: true, regular: true}},
                ...opts,
            }),
        ],
    });

    return result.html;
}

describe('plugin', () => {
    test('should convert vertical tabs to correct new token array', () => {
        // ACT
        const {tokens: result} = makeTransform({content: defaultVerticalContent});

        // ASSERT
        const clearJSON = JSON.parse(JSON.stringify(result.map(({attrs: _, ...item}) => item)));
        expect(clearJSON).toEqual(vertical);
    });

    test('Should convert to correct new token array', () => {
        // ACT
        const {tokens: result} = makeTransform();

        // ASSERT
        const clearJSON = JSON.parse(JSON.stringify(result.map(({attrs: _, ...item}) => item)));
        expect(clearJSON).toEqual(base);
    });

    test('Should use correct attrs', () => {
        // ARRANGE
        const attrs = [
            'data-diplodoc-id',
            'data-diplodoc-key',
            'class',
            'role',
            'aria-controls',
            'aria-selected',
            'tabindex',
            'data-diplodoc-is-active',
        ];

        // ACT
        const {tokens: result} = makeTransform();

        // ASSERT
        const tabs = result.filter(({type}) => type === 'tab_open');
        tabs.forEach((tab, i) => {
            const attrsObject = convertAttrsToObject(tab);

            expect(Object.keys(attrsObject)).toEqual(attrs);
            expect(attrsObject['class']).toEqual(
                `yfm-tab yfm-tab-group${i === 0 ? ' active' : ''}`,
            );
            expect(attrsObject['role']).toEqual('tab');
            expect(attrsObject['tabindex']).toEqual(i === 0 ? '0' : '-1');
        });
    });

    test('Tab should fit tabPanel', () => {
        // ACT
        const {tokens: result} = makeTransform();

        // ASSERT
        const tabs = result.filter(({type}) => type === 'tab_open');
        const tabPanels = result.filter(({type}) => type === 'tab-panel_open');

        expect(tabs.length).toEqual(tabPanels.length);

        tabs.forEach((tab, i) => {
            const tabPanel = tabPanels[i];
            const attrsObject = convertAttrsToObject(tab);
            const panelAttrsObject = convertAttrsToObject(tabPanel);

            expect(attrsObject['aria-controls']).toEqual(panelAttrsObject['id']);
        });
    });

    test('Tab syntax is escaped', () => {
        // ACT
        const {tokens: result} = makeTransform({
            content: ['`{% list tabs %}`'],
        });

        // ASSERT
        expect(result).toEqual(escaped);
    });

    test('Should parse nested tabs', () => {
        // ACT
        const {tokens} = makeTransform({
            content: [
                '{% list tabs %}\n' +
                    '\n' +
                    '- tab1\n' +
                    '\n' +
                    '  content of tab1\n' +
                    '\n' +
                    '  {% list tabs %}\n' +
                    '\n' +
                    '  - nested_tab1\n' +
                    '\n' +
                    '    content of nested tab1\n' +
                    '\n' +
                    '  - nested_tab2\n' +
                    '\n' +
                    '    content of nested tab2\n' +
                    '\n' +
                    '  {% endlist %}\n' +
                    '\n' +
                    '- tab2\n' +
                    '\n' +
                    '  content of tab2\n' +
                    '\n' +
                    '{% endlist %}',
            ],
        });

        // ASSERT
        expect(tokens.map((token) => token.type)).toEqual(nestedTokenTypes);
    });

    it('should return valid token stream if content without indentation', () => {
        // ACT
        const {tokens} = makeTransform({
            content: ['{% list tabs %}', '', '- tab1', '', '> quote', '', '{% endlist %}'],
        });

        // ASSERT
        expect(tokens.map((token) => token.type)).toEqual([
            'tabs_open',
            'tab-list_open',
            'tab_open',
            'inline',
            'tab_close',
            'tab-list_close',
            'tab-panel_open',
            'tab-panel_close',
            'tabs_close',
        ]);
    });

    it('should remove empty tabs', () => {
        // ACT
        const {tokens} = makeTransform({
            content: ['before', '', '{% list tabs %}', '', '', '', '{% endlist %}', '', 'after'],
        });

        // ASSERT
        expect(tokens.map((token) => token.type)).toEqual([
            'paragraph_open',
            'inline',
            'paragraph_close',
            'paragraph_open',
            'inline',
            'paragraph_close',
        ]);
        expect(tokens[1].content).toBe('before');
        expect(tokens[4].content).toBe('after');
    });

    it('should remove tabs with invalid markup inside', () => {
        // ACT
        const {tokens} = makeTransform({
            content: ['{% list tabs %}', '', '> 123', '', '{% endlist %}'],
        });

        // ASSERT
        expect(tokens.map((token) => token.type)).toEqual([]);
    });

    it('should handle simple tab with indents before closing tag', () => {
        const indentsBeforeClosingTag = [
            '# Create a folder',
            '',
            '{% list tabs %}',
            '',
            'aaa',
            '',
            '- Tab with list',
            '',
            'bbb',
            '',
            '   {% endlist %}',
            '',
            'After tabs',
        ];

        const {tokens: result} = makeTransform({content: indentsBeforeClosingTag});

        // ASSERT
        const clearJSON = JSON.parse(JSON.stringify(result.map(({attrs: _, ...item}) => item)));
        expect(clearJSON).toEqual(simpleTab);
    });

    it('should handle complex tab with indents before closing tag', () => {
        const indentsBeforeClosingTag = [
            '# Create a folder',
            '',
            '{% list tabs %}',
            '',
            '- Python',
            '',
            '  About python',
            '',
            '- Tab with list',
            '  - One',
            '  - Two',
            '',
            '- Tab with list',
            '  1. One',
            '  2. Two',
            '',
            '   {% endlist %}',
            '',
            'After tabs',
        ];

        const {tokens: result} = makeTransform({content: indentsBeforeClosingTag});

        // ASSERT
        const clearJSON = JSON.parse(JSON.stringify(result.map(({attrs: _, ...item}) => item)));
        /*
        We finish collecting tokens for generating tabs as soon as we meet the last {% endlist %} tag.
        If it was created with an indent, then in the current example it will be inside the list,
        and after it there will be two additional closing tokens that we remove. This is expected behavior.
         */
        clearJSON[4].map[1] = 16;
        clearJSON[11].map[1] = 16;
        clearJSON.splice(49, 2);

        expect(clearJSON).toEqual(base);
    });

    describe('options', () => {
        test('should add an extra className to container node', () => {
            // ACT
            const {tokens: result} = makeTransform({
                transformOptions: {
                    containerClasses: 'test_1 test_2',
                },
            });

            // ASSERT
            const tabsContainer = result.filter(({type}) => type === 'tabs_open');
            const attrsObject = convertAttrsToObject(tabsContainer[0]);
            expect(attrsObject['class']).toEqual('yfm-tabs test_1 test_2');
        });

        test('should return the default className for container node', () => {
            // ACT
            const {tokens: result} = makeTransform();

            // ASSERT
            const tabsContainer = result.filter(({type}) => type === 'tabs_open');
            const attrsObject = convertAttrsToObject(tabsContainer[0]);
            expect(attrsObject['class']).toEqual('yfm-tabs');
        });

        test('should return custom runtimeJsPath and runtimeCssPath meta data', () => {
            // ACT
            const result = makeTransform({
                transformOptions: {
                    runtimeJsPath: 'path_1',
                    runtimeCssPath: 'path_2',
                },
            });

            // ASSERT
            expect(result.env).toEqual({
                meta: {
                    script: ['path_1'],
                    style: ['path_2'],
                },
            });
        });

        test('should return default runtimeJsPath and runtimeCssPath meta data', () => {
            // ACT
            const result = makeTransform();

            // ASSERT
            expect(result.env).toEqual({
                meta: {
                    script: ['_assets/tabs-extension.js'],
                    style: ['_assets/tabs-extension.css'],
                },
            });
        });
    });

    describe('tabs groups', () => {
        test('should set a random group name for the tabs container', () => {
            // ACT
            const {tokens: result} = makeTransform();

            // ASSERT
            const tabs = result.filter(({type}) => type === 'tabs_open');
            const attrsObject = convertAttrsToObject(tabs[0]);
            expect(attrsObject['data-diplodoc-group']).toMatch(/^defaultTabsGroup-[a-z0-9]{8}$/);
        });

        test('should set a specific group name for the tabs container', () => {
            // ACT
            const {tokens: result} = makeTransform({
                content: [
                    '{% list tabs group=group_1 %}',
                    '',
                    '- Tab with list',
                    '',
                    '{% endlist %}',
                ],
            });

            // ASSERT
            const tabsContainer = result.filter(({type}) => type === 'tabs_open');
            const attrsObject = convertAttrsToObject(tabsContainer[0]);
            expect(attrsObject['data-diplodoc-group']).toEqual('group_1');
        });
    });

    describe('tabs anchors', () => {
        test('should set default anchors for tabs', () => {
            // ACT
            const {tokens: result} = makeTransform();

            // ASSERT
            const tabs = result.filter(({type}) => type === 'tab_open');
            const attrsObject0 = convertAttrsToObject(tabs[0]);
            const attrsObject1 = convertAttrsToObject(tabs[1]);
            const attrsObject2 = convertAttrsToObject(tabs[2]);

            expect(result[result.indexOf(tabs[0]) + 1]?.children?.[0]?.content).toEqual('Python');
            expect(result[result.indexOf(tabs[1]) + 1]?.children?.[0]?.content).toEqual(
                'Tab with list',
            );
            expect(result[result.indexOf(tabs[2]) + 1]?.children?.[0]?.content).toEqual(
                'Tab with list',
            );
            expect(attrsObject0['data-diplodoc-id']).toEqual('python');
            expect(attrsObject1['data-diplodoc-id']).toEqual('tab-with-list');
            expect(attrsObject2['data-diplodoc-id']).toEqual('tab-with-list-1');
            expect(attrsObject0['data-diplodoc-key']).toEqual('python');
            expect(attrsObject1['data-diplodoc-key']).toEqual('tab%20with%20list');
            expect(attrsObject2['data-diplodoc-key']).toEqual('tab%20with%20list');
            expect(attrsObject0['data-diplodoc-is-active']).toEqual('true');
            expect(attrsObject1['data-diplodoc-is-active']).toEqual('false');
            expect(attrsObject2['data-diplodoc-is-active']).toEqual('false');
        });

        test('should set custom anchors for tabs', () => {
            // ACT
            const {tokens: result} = makeTransform({
                content: [
                    '{% list tabs %}',
                    '',
                    '- Python {#my-tab}',
                    '',
                    '- Java {#my-tab}',
                    '',
                    '{% endlist %}',
                ],
            });

            // ASSERT
            const tabs = result.filter(({type}) => type === 'tab_open');

            const attrsObject0 = convertAttrsToObject(tabs[0]);
            const attrsObject1 = convertAttrsToObject(tabs[1]);

            expect(result[result.indexOf(tabs[0]) + 1]?.children?.[0]?.content).toEqual('Python');
            expect(result[result.indexOf(tabs[1]) + 1]?.children?.[0]?.content).toEqual('Java');
            expect(attrsObject0['data-diplodoc-id']).toEqual('my-tab');
            expect(attrsObject1['data-diplodoc-id']).toEqual('my-tab-1');
            expect(attrsObject0['data-diplodoc-key']).toEqual('my-tab');
            expect(attrsObject1['data-diplodoc-key']).toEqual('my-tab');
        });

        test('should process tabs with special letters', () => {
            // ACT
            const {tokens: result} = makeTransform({
                content: [
                    '{% list tabs %}',
                    '',
                    '',
                    '- C#',
                    '',
                    '',
                    '- C++',
                    '',
                    '',
                    '{% endlist %}',
                ],
            });

            // ASSERT
            const tabs = result.filter(({type}) => type === 'tab_open');
            const attrsObject0 = convertAttrsToObject(tabs[0]);
            const attrsObject1 = convertAttrsToObject(tabs[1]);

            expect(result[result.indexOf(tabs[0]) + 1]?.children?.[0]?.content).toEqual('C#');
            expect(result[result.indexOf(tabs[1]) + 1]?.children?.[0]?.content).toEqual('C++');
            expect(attrsObject0['data-diplodoc-id']).toEqual('c');
            expect(attrsObject1['data-diplodoc-id']).toEqual('c-1');
            expect(attrsObject0['data-diplodoc-key']).toEqual('c%23');
            expect(attrsObject1['data-diplodoc-key']).toEqual('c%2b%2b');
        });
    });

    describe('html snapshots', () => {
        beforeAll(() => {
            let i = 0;
            const values = [
                0.123456789, 0.987654321, 0.678912345, 0.214365879, 0.456789123, 0.123789456,
            ];

            jest.spyOn(global.Math, 'random').mockImplementation(() => values[i++ % values.length]);
        });

        afterAll(() => {
            jest.spyOn(global.Math, 'random').mockRestore();
        });

        it('should render common tabs', () => {
            expect(
                html(
                    dedent`
                    {% list tabs %}

                    - Tab 1

                      Content of tab 1

                    - Tab 2

                      Content of tab 2

                    {% endlist %}
                `,
                ),
            ).toMatchSnapshot();
        });

        it('should render radio tabs', () => {
            expect(
                html(
                    dedent`
                    {% list tabs radio %}

                    - Radio 1

                      Content of radio 1

                    - Radio 2

                      Content of radio 2

                    {% endlist %}
                `,
                ),
            ).toMatchSnapshot();
        });
    });
});
