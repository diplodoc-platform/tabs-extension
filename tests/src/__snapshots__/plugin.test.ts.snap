// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin html snapshots should render common tabs 1`] = `
<div class="yfm-tabs"
     data-diplodoc-group="defaultTabsGroup-4fzzzxjy"
     data-diplodoc-variant="regular"
>
  <div class="yfm-tab-list"
       role="tablist"
  >
    <div data-diplodoc-id="tab-1"
         data-diplodoc-key="tab%201"
         class="yfm-tab yfm-tab-group active"
         role="tab"
         aria-controls="zk00000y"
         aria-selected="true"
         tabindex="0"
         data-diplodoc-is-active="true"
    >
      Tab 1
    </div>
    <div data-diplodoc-id="tab-2"
         data-diplodoc-key="tab%202"
         class="yfm-tab yfm-tab-group"
         role="tab"
         aria-controls="ofvc1cad"
         aria-selected="false"
         tabindex="-1"
         data-diplodoc-is-active="false"
    >
      Tab 2
    </div>
  </div>
  <div id="zk00000y"
       class="yfm-tab-panel active"
       role="tabpanel"
       aria-labelledby="tab-1"
       data-title="Tab 1"
  >
    <p>
      Content of tab 1
    </p>
  </div>
  <div id="ofvc1cad"
       class="yfm-tab-panel"
       role="tabpanel"
       aria-labelledby="tab-2"
       data-title="Tab 2"
  >
    <p>
      Content of tab 2
    </p>
  </div>
</div>
`;

exports[`plugin html snapshots should render radio tabs 1`] = `
<div class="yfm-tabs yfm-tabs-vertical"
     data-diplodoc-group="defaultTabsGroup-7ptgcyuj"
     data-diplodoc-variant="radio"
>
  <div data-diplodoc-id="radio-1"
       data-diplodoc-key="radio%201"
       class="yfm-tab yfm-tab-group yfm-vertical-tab"
       role="tab"
       aria-controls="gfzybi81"
       aria-selected="false"
       tabindex="0"
       data-diplodoc-is-active="false"
       data-diplodoc-vertical-tab="true"
  >
    <input class="radio"
           type="radio"
    >
    <label>
      Radio 1
    </label>
  </div>
  <div id="gfzybi81"
       class="yfm-tab-panel"
       role="tabpanel"
       aria-labelledby="radio-1"
       data-title="Radio 1"
  >
    <p>
      Content of radio 1
    </p>
  </div>
  <div data-diplodoc-id="radio-2"
       data-diplodoc-key="radio%202"
       class="yfm-tab yfm-tab-group yfm-vertical-tab"
       role="tab"
       aria-controls="4gfir17c"
       aria-selected="false"
       tabindex="-1"
       data-diplodoc-is-active="false"
       data-diplodoc-vertical-tab="true"
  >
    <input class="radio"
           type="radio"
    >
    <label>
      Radio 2
    </label>
  </div>
  <div id="4gfir17c"
       class="yfm-tab-panel"
       role="tabpanel"
       aria-labelledby="radio-2"
       data-title="Radio 2"
  >
    <p>
      Content of radio 2
    </p>
  </div>
</div>
`;
